clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java

```diff
@@ -1871,6 +1881,138 @@ private void testSplitBatchAndSend(TransactionManager txnManager,
         }
     }
 
+    @Test
+    public void testNoDoubleDeallocation() throws Exception {
+        long deliverTimeoutMs = 1500L;
+        long totalSize = 1024 * 1024;
+        String metricGrpName = "producer-custom-metrics";
+        MatchingBufferPool pool = new MatchingBufferPool(totalSize, batchSize, metrics, time, metricGrpName);
+        setupWithTransactionState(null, false, pool);
+
+        // Send first ProduceRequest
+        Future<RecordMetadata> request1 =
+            accumulator.append(tp0, time.milliseconds(), "key".getBytes(), "value".getBytes(), null, null, MAX_BLOCK_TIMEOUT).future;
+        sender.run(time.milliseconds());  // send request
+        assertEquals(1, client.inFlightRequestCount());
+        time.sleep(deliverTimeoutMs);
+        sender.run(time.milliseconds());  // expire the batch
+        assertTrue(request1.isDone());
+        // make sure the batch got deallocated right after expiry
+        assertTrue(pool.allMatch());
```

Comment:
How is this testing that we don't double deallocate. Ideally, we would expire the inflight batch, then get a response, and then check that the batch is not deallocated twice. In this case, it would de allocate only once, unless I am missing something.

Reply:
updated the test a bit to add another `sender.run(time.milliseconds());` call after  the batch expiry. Not sure if that fits `expire the inflight batch, then get a response`. 

my expectation was that if there is double deallocation, we would get an `IllegalStateException` exception from  `MatchingBufferPool.deallocate`.